.circleci/config.yml

```diff
@@ -16,8 +32,72 @@ jobs:
           path: obj/release/mode-voltdb-community.tar.gz
           destination: mode-voltdb-community.tar.gz
 
-workflows:
-  version: 2
-  build_and_package:
-    jobs:
-      - dist
+
+  build-docker-image:
+    resource_class: xlarge
+    docker:
+      - image: ubuntu:16.04
+    steps:
+      - setup_docker_image_context:
+          docker_login_username: $MODECI_GITHUB_USERNAME
+          docker_login_token: $MODECI_GITHUB_USER_TOKEN
+      - run:
+          name: Build and publish docker image
+          command: |
+            DOCKER_TAG=`cat version.txt`
+            docker build \
+                -f .circleci/docker/Dockerfile \
+                -t docker.pkg.github.com/mode/voltdb/voltdb:${DOCKER_TAG}-${CIRCLE_SHA1} \
+                .
+            docker push docker.pkg.github.com/mode/voltdb/voltdb:${DOCKER_TAG}-${CIRCLE_SHA1}
+
+
+  publish-docker-image:
+    resource_class: xlarge
+    docker:
+      - image: ubuntu:16.04
+    steps:
+      - setup_docker_image_context:
+          docker_login_username: $MODECI_GITHUB_USERNAME
+          docker_login_token: $MODECI_GITHUB_USER_TOKEN
+      - run:
+          name: Build and publish docker image
+          command: |
+            DOCKER_TAG=`cat version.txt`
+            docker build -f .circleci/docker/Dockerfile \
+                -t docker.pkg.github.com/mode/voltdb/voltdb:${DOCKER_TAG}-${CIRCLE_SHA1} \
+                -t docker.pkg.github.com/mode/voltdb/voltdb:${DOCKER_TAG} \
+                -t docker.pkg.github.com/mode/voltdb/voltdb:latest \
+                .
+            docker push docker.pkg.github.com/mode/voltdb/voltdb:${DOCKER_TAG}-${CIRCLE_SHA1}
+            docker push docker.pkg.github.com/mode/voltdb/voltdb:${DOCKER_TAG}
+            docker push docker.pkg.github.com/mode/voltdb/voltdb:latest
+
+
+
+commands:
+
+  setup_docker_image_context:
+    description: "Setup docker image builder/publisher container"
+    parameters:
+      docker_login_username:
+        type: string
+      docker_login_token:
+        type: string
+    steps:
+      - run: apt-get update
+      - run: apt-get install -y git
+      - run:
+          name: Install Docker CLI Dependencies
+          command: |
+            apt-get update
+            apt-get install -y curl apt-transport-https ca-certificates gnupg-agent software-properties-common
+            curl -fsSL https://download.docker.com/linux/ubuntu/gpg | apt-key add -
+            apt-key fingerprint 0EBFCD88
+            add-apt-repository "deb [arch=amd64] https://download.docker.com/linux/ubuntu $(lsb_release -cs) stable"
+            apt-get update
+            apt-get install docker-ce-cli
```

Comment:
Any reason not to build these commands into the base image we use for this build?

Reply:
Do you mind elaborating? I'm not sure what you mean. Which base image?

Reply:
The image that we use to run the the build on circle. Here we are starting from `ubuntu:16.04`, but I'm wondering if we can pre-build an image with these already run (since they don't change). Then use that image for these builds to save time. But I may be misunderstanding when all these commands get run

Reply:
Ah, that makes sense. That seems like a reasonable change to make. But, could it also be a future iteration potentially?

Reply:
These steps don't actually take too much either from the build. The VoltDB build is the monster here haha.

Reply:
Re-running the `apt` commands each time also ensures we don't have a stale base image. A stale base image is going to be prone to security issues, which will necessitate doing emergency rebuilds for. This way we lean on the apt repositories to always be up to date with regards to security vulnerabilities.

Reply:
That's a good point regarding stale images.

Reply:
Both points seem reasonable to me, SGTM